htlcswitch/link.go

```diff
@@ -2506,8 +2506,22 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 			}
 		}
 
+		// If we can send updates then we can process adds in case we
+		// are the exit hop and need to send back resolutions, or in
+		// case there are validity issues with the packets. Otherwise
+		// we defer the action until resume.
+		//
+		// We are free to process the settles and fails without this
+		// check since processing those can't result in further updates
+		// to this channel link.
+		if l.quiescer.canSendUpdates() {
+			l.processRemoteAdds(fwdPkg)
+		} else {
+			l.quiescer.onResume(func() {
+				l.processRemoteAdds(fwdPkg)
```

Comment:
Will this override each time? Or cause us to try to process many times once stfu is fin?

Reply:
It does not override it queues. This can't actually happen more than once though.

The reason it can't happen more than once is as follows:
1. stfu must wait until after all updates from the stfu sender are committed on both commitment transactions
2. my updates are committed to the remote after I send commit_sig
3. my updates are committed to the local after I receive R&A and then commit_sig
4. Now I can send stfu
5. Now I send R&A <-- this is the culprit for this case

Now if we take it from the remotes perspective, I receive that R&A described in step 5 above, which causes me to take all those HTLCs and then process them in earnest. Ordinarily I'd recommend we just forward them rather than deferring, however, there's an issue because the exit hop processing is done in the link instead of the switch (which I continue to believe is something we should rectify), the exit hop processing code notices any HTLCs that terminate with us and then immediately tries to resolve them with a fulfill message, which causes us to commit a protocol violation since the channel is quiescent.

Since the channel is quiescent by this point, we cannot receive any more updates from the remote and therefore we won't do any more commit dances meaning this will not trigger more than once. We have other code that handles those protocol violations so at this point it seems fine to defer the processing of this package until quiescence resumes.

Now, as I type this I'm realizing there's a tricky situation that may unfold if the node crashes between the time that we defer the processing of this forward package, and the time we actually process it. IIRC we still re-process the forward package on link restart if it hasn't been processed yet. Since we don't process it, we don't populate the ack filter and so when we reload the forwarding package it should work? I'd appreciate feedback on this claim though.

Reply:
@Crypt-iQ care to elaborate?